### ISSUE TITLE
EnforcedShorthandSyntax: Add option to allow either syntax but enforce consistency within a single hash

### ISSUE THREAD

**Issue by hpello** [rubocop/rubocop#11436]
## Is your feature request related to a problem? Please describe.

💡 In a big codebase, we would like to gradually start using Hash Shorthand Syntax (3.1).

Ideally, we would like to be able to:
1. use it in **new** lines of code
2. do not use it in **old** lines of code (for now)
3. enforce consistency of the syntax within a single hash (either all shorthand, or all explicit hash literal)

👎 We are not able to achieve this with the current options of the `EnforcedShorthandSyntax` option:
- `always` and `consistent` complain on existing code (which fully uses explicit hash literal syntax)
- `never` does not allow the shorthand syntax
- `either` allows it, but will not enforce **consistency** within a single hash

This is preventing us from using this cool new syntactic sugar feature!

## Describe the solution you'd like

We would like an ⭐️ Additional Style ⭐️ for the `EnforcedShorthandSyntax` option that allows:
- **both** syntaxes for hashes of a codebase
- **only one** syntax to be valid within a hash

Examples:
```ruby
# EnforcedShorthandSyntax: always
foo: foo            # => ❌ error: omit the hash for foo
foo: foo, lol: true # => ❌ error: omit the hash for foo
foo:, lol: true     # => ✅

# EnforcedShorthandSyntax: consistent
foo: foo            # => ❌ error: omit the hash for foo
foo: foo, lol: true # => ✅
foo:, lol: true     # => ❌ error: Do not mix explicit and implicit hash values. Include the hash value.
```
```ruby
# EnforcedShorthandSyntax: ???
foo: foo            # => ✅
foo: foo, lol: true # => ✅
foo:, lol: true     # => ❌ error: Do not mix explicit and implicit hash values. Include hash values everywhere or nowhere.
```

## Describe alternatives you've considered

- Forbiding hash shorthand syntax 🙁 (using `never`)
- Allowing it with `either`, but without consistency within a hash 🙁

## Additional context

For what it's worth, a similar shorthand syntax exists in JavaScript. ESLint has a [rule and options](https://eslint.org/docs/latest/rules/object-shorthand#options) that allow for fine tuning in consistency.

Here is a table that shows available options and their equivalent in RuboCop:

| ESLint `object-shorthand` | Rubocop `EnforceShorthandSyntax` |
| ----------- | ----------- |
| `always`       | `always`       |
| `never`       | `never`       |
| `either`       | `either`       |
| `consistent`       | _this feature request_       |
| `consistent-as-needed`       | `consistent`       |

### Naming suggestions for new style
1. `consistent-either`: short and self-explanatory
1. `consistently-either`: to emphasis that this is _consistent_ version of the `either` rule, but sounds weird to me
1. `consistent-without-preference`, `consistent-no-preference`, `consistent-impartial`: to show that it is an unbiased version of the `consistent` rule
1. `evenly-consistent`, `equally-consistent`: to insist on the fact that the `consistency` is enforced _evenly_ for both syntaxes
1. ... ?

Do you think it could be a nice addition? 😃
If so, what would be the recommended way to implement it?

Regards,

**Comment by magni-** [comment]
Have you looked into using https://github.com/prontolabs/pronto-rubocop ? You'd configure `rubocop` to enforce the desired style moving forward, but then `rubocop` would only run against changed code—so old, untouched code wouldn't cause your linting check to fail.

**Comment by hpello** [comment]
Hi! Thanks for the suggestion.
Unfortunately this is not suited for our usage: that would be a blocker to the occasional/refactoring that happens in the codebase.

**Comment by hpello** [comment]
Hi! Thanks for the suggestion.
Unfortunately this is not suited for our usage: that would be a blocker to the occasional/refactoring that happens in the codebase.

**Comment by magni-** [comment]
`pronto` runs against changed _lines_, not changed files. If you're touching a line as part of a refactor, wouldn't you then update it to the new syntax as part of that refactoring?

**Comment by dvandersluis** [comment]
You can use `EnforcedStyle: either_consistent` now, so I'll close this.

**Comment by hpello** [comment]
You mean `consistent_either`? 🙂

Thanks a lot for implementing this!
